EIPS/eip-7979.md

```diff
@@ -0,0 +1,435 @@
+
+Even offline, many static analyses can become impractically slow in depth and at scale.  This include verifying code against formal specifications and other proofs of program invariants —  with so much at stake on the blockchain correct contracts are crucial.  These analyses must first get past the difficulties of tracing control flow (_Analyzing Smart Contracts: From EVM to a sound Control-Flow Graph._) and one proposal goes so far as using neural networks to recover structure from code, _Neural-FEBI: Accurate Function Identification in Ethereum Virtual Machine Bytecode._  This proposal renders these problems trivial.
+
+#### Taming the EVM
+
+To prevent tracing from "going quadratic" we must prevent the dynamic use of jumps.  Most JUMP and JUMPI instructions in EVM code are preceded by a `PUSH` — that is, they are static.  So this proposal can validate that `JUMP` and `JUMPI` are used statically.   The only places that jumps _must_ be used dynamically are to support calls to Turing's _subsidiary operations._  For that purpose we propose `JUMPSUB` and `RETURNSUB` opcodes as replacements. 
+
+## Specification
+
+### `JUMPSUB (0x..)`
+
+Transfers control to a subsidiary operation.
+
+1. Pop the `destination` from the `data stack`.
+2. Push the current `PC + 1` to the `return stack`.
+3. Set `PC` to `destination`.
+
+The gas cost is _mid_ (`8`).
+
+### `RETURNSUB (0x..)`
+
+Returns control to the caller of a subsidiary operation.
+
+1. Pop the `return stack` to `PC`.
+
+The gas cost is _verylow_ (`3`).
+
+### `MAGIC (0xEF....)`
+
+After this EIP has been activated code beginning with the `MAGIC` bytes MUST be the leading bytes of a `valid` program.  Execution begins immediately after the MAGIC bytes.  (Opcode and magic values are still to be determined.)
+
+_Notes:_
+
+* _Values popped off the `return stack` do not need to be validated, since they are alterable only by `JUMPSUB` and `RETURNSUB`._
+* _The description above lays out the semantics of these instructions in terms of a `return stack`.  But the actual state of the `return stack` is not observable by EVM code or consensus-critical to the protocol.  (For example, a node implementer may code `JUMPSUB` to unobservably push `PC` on the `return stack` rather than `PC + 1`, which is allowed so long as `RETURNSUB` observably returns control to the `PC + 1` location.)_
+
+### Validity
+
+Code beginning with `MAGIC` MUST be `valid` Conditions on `valid` code MUST be validated statically, at `CREATE` time, in time and space linear in the size of the code.
+
+#### Constraints on `valid` EVM code
+
+* Every executable instruction MUST be `valid`:
+  * All opcode values must be `valid` -- they must have been defined in the Yellow Ppaer or an EIP and not since deprecated.  The INVALID opcode is defined.
```

Comment:
With deprecated opcodes I assume you mean CALLCODE and SELFDESTRUCT, but I'm not sure if these are "officially" deprecated, for instance the deprecate CALLCODE EIP is stagnant https://eips.ethereum.org/EIPS/eip-2488. I'd also mention that INVALID is `0xfe`, because all other opcodes which are not defined map to `INVALID` (but they should not be used as such since these opcodes could be defined in a new fork)